File: day9/p1.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

#define ll unsigned long long

ll solve(vector<ll> fs)
{
    int left = 0;
    int right = fs.size() - 1;
    while (right >= left)
    {
        if (fs[left] == -1)
        {
            swap(fs[left], fs[right]);
            while (fs[right] == -1)
            {
                --right;
            }
        }
        ++left;
    }

    ll total = 0;
    for (ll i = 0; i < fs.size(); ++i)
    {
        if (fs[i] != -1)
        {
            total += i * fs[i];
        }
    }
    return total;
}

int main()
{
    string data;
    cin >> data;
    vector<ll> fs;
    for (ll i = 0; i < data.size(); ++i)
    {
        int d = data[i] - '0';
        for (int j = 0; j < d; ++j)
        {
            fs.push_back(i % 2 == 0 ? i / 2 : -1);
        }
    }
    cout << solve(fs) << endl;
    return 0;
}

File: day9/p1.py
def solve(fs):
    left = 0
    right = len(fs) - 1

    while right >= left:
        if fs[left] == -1:
            fs[left], fs[right] = fs[right], fs[left]  # Python swap
            while fs[right] == -1:
                right -= 1
        left += 1

    total = 0
    for i in range(len(fs)):
        if fs[i] != -1:
            total += i * fs[i]
            if i * fs[i] < 0:
                print(f"Error: {i} {fs[i]}")

    return total


def main():
    data = input()
    fs = []

    for i in range(len(data)):
        d = int(data[i])
        for j in range(d):
            fs.append(i // 2 if i % 2 == 0 else -1)

    print(solve(fs))


if __name__ == "__main__":
    main()
